### PR TITLE
Boundary slice excluded

### PR DESCRIPTION
**Pull request by TomAugspurger**
Changes `boundary_slice` to handle cases where

- the index is not sorted
- using label-based indexing (loc)
- the start or stop is missing

See https://github.com/pandas-dev/pandas/issues/8613 for details on the pandas
side, but the short version is pandas will raise a KeyError doing `.loc[start:stop]` if the index is not monotonic and start or stop are missing.

Closes https://github.com/dask/dask/issues/2211

### CODE REVIEW

File: dask/dataframe/methods.py [hunk 1]
```diff
@@ -55,7 +55,18 @@ def boundary_slice(df, start, stop, right_boundary=True, left_boundary=True,
     2  20
     2  30
     """
-    result = getattr(df, kind)[start:stop]
+    if kind == 'loc' and not df.index.is_monotonic:
+        # Pandas treats missing keys differently for label-slicing
+        # on monotonic vs. non-monotonic indexes
+        # If the index is monotonic, `df.loc[start:stop]` is fine.
+        # If it's not, `df.loc[start:stop]` raises when `start` is missing
+        if start:
```
Comment (mrocklin) [inline dask/dataframe/methods.py]: `if start is not None`?

Comment (TomAugspurger) [inline dask/dataframe/methods.py]: Yeah, you're right; fixed.

**Comment by mrocklin** [comment]
Thanks @TomAugspurger !

@shughes-uk I *think* that this should resolve the issues that you were seeing.

**Comment by shughes-uk** [comment]
I still get a keyerror when running the test I posted in the issue thread.

**Pull request merged**